grape/general_graph.py

```diff
@@ -561,20 +400,8 @@ def local_efficiency(self):
             It is in the range [0, 1].
         """
 
-        for node in self:
-            subgraph = list(self.successors(node))
-            denom_subg = len(list(subgraph))
-
-            if denom_subg != 0:
-                sum_efficiencies = 0
-                for w in list(subgraph):
-                    kv_efficiency = self.nodes[w]["nodal_eff"]
-                    sum_efficiencies = sum_efficiencies + kv_efficiency
-
-                self.nodes[node]["local_eff"] = sum_efficiencies / denom_subg
-
-            else:
-                self.nodes[node]["local_eff"] = 0.
+        loceff = self.local_efficiency_kernel(list(self))
+        nx.set_node_attributes(self, loceff, name="local_eff")
 
     def global_efficiency(self):
```

Comment:
I do not understand: if `global_efficiency` is a graph attribute, I wouldn't store such quantity in all the nodes, it looks extremely redundant. If I'm not wrong, it should be a class property!

Reply:
I have made it a `property`, it should correctly be so.
I have a question/idea. For now, I still have the check on `nodal_eff`. Ideally, if the graph has already been loaded, to _get_ `global_efficiency`, we should follow these steps:
 ```
g.calculate_shortest_path()                                                 
g.nodal_efficiency() 
```
this is similar for other measures. If we encapsulate these instructions within a function, the user should not be aware of how to extract attributes.
However, when we need to compute more than one measure, this would mean recalculating `g.calculate_shortest_path()` and  `g.nodal_efficiency` for every attribute qe invoke. Is there a smart way to solve this matter? Tks

Reply:
Great, that is the (part of the) design I would to converge to. My idea was something like:
```
efficiency = nx.get_node_attributes('eff')
if efficiency: return efficiency

# only if eff is not already stored within the nodes
eff = compute_eff() # the current implementation here
set_node_attributes(eff)
```
Of course for the different measures. In this case, if the attribute is set we neglect the computation and just return it, otherwise we perform the measure. Do this sound fine for you?

Reply:
It sounds ok, the only concern that I have is that when the graph gets damaged, we _need_ to recompute. I would naively add a parameter to `efficiency`, something like `force_recomputation` (or any other name that we may like). Is this the way to do so, or is it there a more elegant/pythonic way to do so?

Reply:
We can add an optional argument (avoiding the `property` decorator) but it still remains a problem in the user-side: in the typical non-developer workflow, the user simulates a fault on the system and calls the measures. With an optional argument, in this cases it is really easy to insert a bug. I would prefer to remove the nodes attributes in the damage procedure, making less prone to inconsistent states.
PS Take into account that we have similar problem adding manually a node (for example after the `load`).

Reply:
I see what you say; the additional argument was just an idea.
What I do not see how the measures are updated whenever the graph changes (for the removal of one node or, as you pointed out correctly, the addition of one) if we do not implement the recalculation of the shortest paths - and of other measures on which the attribute depends - every time we call the attribute itself.
What if we want to know, for example, the efficiency after the damage? The dictionary `efficiency` is still present, and the code `if efficiency: return efficiency` will work, just the `efficiency` is not the right one because the shortest path will have changed in the meantime due to the damage.

Reply:
Yes, but everything should be solved by simply adding a function like:
```
def clear_data(self, attributes_to_remove): # a list of string with all the attributes we want to remove
  for nodes in self: [del nodes[attr] for attr in attibutes_to_remove]
```
and using this wherever we need (adding nodes, removing nodes, breaking something, ...). The user does not see anything of this, and we are pretty sure to do recalculation only when we really need.

Reply:
Just one brief comment: there are two attributes that I think could still be turned into properties (namely, `IntermediateStatus` and `FinalStatus`), but since we use them just in `FaultDiagnosis()` class I decided to correct them when we arrive at the analysis of that class, together with the other modifications that will be needed.